### ISSUE TITLE
nixos-rebuild cannot be made to build locally with a remote target

### ISSUE THREAD

**Issue by robryk** [NixOS/nixpkgs#211459]
### Describe the bug

There's no way to get nixos-rebuild to build locally with a nonempty `--target-host`: if `--target-host` is set but `--build-host` is unset, building happens on target host. In the past specifying `--build-host localhost` would have had that effect, but that's not the case anymore since cbd1dc1e725c4661c0ee556b8339b3c50d8a934c (cc @l0b0 @veprbl).

### Steps To Reproduce
1. Run `nixos-rebuild --target-host somehost switch`
2. Observe that building happens on target host instead of locally.
3. Observe that documentation says that empty `--build-host` means that building happens locally.

### Expected behavior
First and foremost, there should be *a* way to get stuff built locally with a remote target host.

Secondly, the documentation should correctly describe where building happens for different values of `--build-host`.

### Notify maintainers

@Profpatsch as maintainer of nixos-rebuild

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.80, NixOS, 23.05 (Stoat), 23.05.git.27e99b84dde`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.12.0`
 - channels(root): `""`
 - channels(robryk): `""`
 - nixpkgs: `/etc/nixpkgs`
```

**Comment by l0b0** [comment]
@robryk If I understand correctly, `--build-host localhost` should result in the build happening on the host called localhost, which very likely is the local host. The only overhead should be an SSH connection to localhost, which should be tiny.

**Comment by robryk** [comment]
This assumes that ssh is actually setup on localhost, and that current user can ssh to localhost without jumping through any hoops (e.g. confirming that they really want to use the key from the agent).

Also, I expect that doing so will create amusingly confusing invocations of `nix-copy-closure

**Comment by l0b0** [comment]
Good points, and there definitely should be a way to force building on the local host. Hopefully it can be achieved without reverting to the original behaviour, since that was both confusing (why was only "localhost" special, and not "127.0.0.1", "::1", "localhost.localdomain", "foo.local", etc?) and technically incorrect ([the "localhost" string does not necessarily mean the local host](https://github.com/NixOS/nixpkgs/issues/148918#issuecomment-1124422265)).

Maybe a `--local-build` or similar flag would be a good way to avoid all three issues?

**Comment by robryk** [comment]
I would frankly expect (and documentation implies that) that _not_ specifying `--build-host` would cause the build to happen locally (i.e. on the machine where `nixos-rebuild` is run). Do you know why that's not the implemented behaviour?

**Comment by l0b0** [comment]
No idea, sorry. Tracing the history of [the relevant lines](https://github.com/NixOS/nixpkgs/blob/cbd1dc1e725c4661c0ee556b8339b3c50d8a934c/pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh#L151-L153) backwards:

1. [fb2eae08d53606823051dead0d07b6461e82ec1b](https://github.com/NixOS/nixpkgs/commit/fb2eae08d53606823051dead0d07b6461e82ec1b#diff-4efef885a98a7c3dcdc7c4cb4fcaa10480fdc0da3627afaa8f5513edc068ab35R94-R96)
2. [78be7f5a53e3945b72482a99c639f5fe8cc75fd6](https://github.com/NixOS/nixpkgs/commit/78be7f5a53e3945b72482a99c639f5fe8cc75fd6#diff-4efef885a98a7c3dcdc7c4cb4fcaa10480fdc0da3627afaa8f5513edc068ab35L94-L96)
3. [ca0c46040649ab4a6d5d7bc838b393fdcbcae10c](https://github.com/NixOS/nixpkgs/commit/ca0c46040649ab4a6d5d7bc838b393fdcbcae10c#diff-4efef885a98a7c3dcdc7c4cb4fcaa10480fdc0da3627afaa8f5513edc068ab35R94-R96) finally provides an explanation:

   > As a convenience, if only --target-host is specified, --build-host is
implicitly set to that host too. So if you want to build locally and deploy
remotely you have to explicitly set "--build-host localhost".

**Comment by robryk** [comment]
ISTM that this is really undesirable magic. Perhaps we could:
1. Cause lack of `--build-host` to always cause builds to happen locally.
2. Temporarily cause both `--build-host` set to `localhost` as well as `--build-host` unset when `--target-host` is set to cause a warning about recently changed semantics to be printed.

Would you consider this reasonable?

**Comment by l0b0** [comment]
That looks like an excellent way forward. The message might be completely drowned in other output, but at least it would be _possible_ to figure out why behaviour has changed.

**Comment by robryk** [comment]
Great. I will make a PR to that effect ~tomorrow.

**Comment by zendo** [comment]
https://github.com/NixOS/nixpkgs/pull/210382

**Comment by robryk** [comment]
This was fixed by #210382.
